### PR TITLE
Add A/B testing for GraphQL on supported document types

### PR DESCRIPTION
**Pull request by brucebolt**
, [Jira issue PP-2556](https://gov-uk.atlassian.net/browse/PP-2556)This will allow us to serve a defined amount of traffic from GraphQL, to measure response times in a real environment.

Note: all requests where the GraphQL variant is in use will involve a query to Publishing API's GraphQL endpoint, since we need to determine whether the document is a news article before deciding how to render it.  This should give us a good indication of how our implementation would cope with all documents types being migrated to use GraphQL.

[Trello card](https://trello.com/c/stxnESVI)

### CODE REVIEW

File: lib/content_item_loader.rb [hunk 1]
```diff
@@ -43,7 +43,22 @@ def load_from_sources(base_path)
...
 
   def use_graphql?
-    request&.params&.[]("graphql") == "true"
+    if request && request.params["graphql"] == "true"
+      return true
+    elsif request && request.params["graphql"] == "false"
+      return false
+    end
+
+    if request && request.headers
```
Comment (mike3985) [inline lib/content_item_loader.rb:52]: We're checking that `request` isn't nil for every conditional in this method. It might be nice to just check that once at the beginning of the method?

File: lib/content_item_loader.rb [hunk 2]
```diff
@@ -43,7 +43,22 @@ def load_from_sources(base_path)
...
+      ab_test = GovukAbTesting::AbTest.new(
+        "GraphQLRoles",
+        allowed_variants: %w[A B Z],
+        control_variant: "Z",
+      )
+      @requested_variant = ab_test.requested_variant(request.headers)
+
+      true if @requested_variant.variant?("B")
+    end
```
Comment (mike3985) [inline lib/content_item_loader.rb:61]: With these changes, there are scenarios where this method can now return nil instead of just true or false

File: spec/unit/content_item_loader_spec.rb [hunk 3]
```diff
@@ -80,16 +203,90 @@
...
+        end
+      end
+    end
+
+    context "with ALLOW_LOCAL_CONTENT_ITEM_OVERRIDE=true" do
+      before do
+        ENV["ALLOW_LOCAL_CONTENT_ITEM_OVERRIDE"] = "true"
+        stub_const("ContentItemLoader::LOCAL_ITEMS_PATH", "spec/fixtures/local-content-items")
+      end
+
```
Comment (mike3985) [inline spec/unit/content_item_loader_spec.rb:271]: It looks like the additions below this line are accidental duplications. Unless I'm failing a spot the difference test 😵‍💫

File: lib/content_item_loader.rb [hunk 4]
```diff
@@ -43,7 +43,22 @@ def load_from_sources(base_path)
...
-    request&.params&.[]("graphql") == "true"
+    if request && request.params["graphql"] == "true"
+      return true
+    elsif request && request.params["graphql"] == "false"
+      return false
+    end
+
+    if request && request.headers
+      ab_test = GovukAbTesting::AbTest.new(
+        "GraphQLRoles",
```
Comment (mike3985) [inline lib/content_item_loader.rb:54]: I was expecting Frontend to be handling the News Articles A/B test and Collections to be handling Roles?

Comment (mike3985) [reply lib/content_item_loader.rb:54]: I just double-checked for my own sanity... Collections is already handling the Roles test. I guess this is just a copy-paste error 😄

File: lib/content_item_loader.rb (hunk 1 above)
Comment (brucebolt) [reply lib/content_item_loader.rb:52]: I hadn't spotted that.  I'll add a guard clause at the beginning of the method.

File: lib/content_item_loader.rb (hunk 2 above)
Comment (brucebolt) [reply lib/content_item_loader.rb:61]: I think that would've been the case before it there's no `request`, since the safe nav operator will return nil.

File: lib/content_item_loader.rb (hunk 4 above)
Comment (brucebolt) [reply lib/content_item_loader.rb:54]: Oops.  I'll update this to `GraphQLNewsArticles`.

File: lib/content_item_loader.rb (hunk 2 above)
Comment (mike3985) [reply lib/content_item_loader.rb:63]: I still think the previous version was doing the right thing because it would always ultimately compare the result of that safe-navigated-to value with `"true"`

Comment (mike3985) [reply lib/content_item_loader.rb:63]: I think it's good form for a Ruby `?` predicate method to produce booleans in general?

Comment (brucebolt) [reply lib/content_item_loader.rb:63]: I've updated to always return a boolean value, including an explicit `return false` for when none of the conditions have been met.

**Review by mike3985 — APPROVED** [review]

File: lib/content_item_loader.rb [hunk 5]
```diff
@@ -43,7 +43,26 @@ def load_from_sources(base_path)
...
-    request&.params&.[]("graphql") == "true"
+    return false unless request
+
+    if request.params["graphql"] == "true"
+      return true
+    elsif request.params["graphql"] == "false"
+      return false
+    end
+
+    if request.headers
```
Comment (KludgeKML) [inline lib/content_item_loader.rb:54]: There isn't a test for if request.headers is nil (we also appear to have lost coverage of line 36 above, testing if there's a graphql response, but it's not in the allowed schemas?

Comment (brucebolt) [reply lib/content_item_loader.rb:54]: I've added a test for headers being `nil` and also ensured we're testing on the B variant for schemas that aren't allowed.

Comment (KludgeKML) [reply lib/content_item_loader.rb:54]: Hmm - this line is still showing as only being covered for the case when request.headers is not nil. (I can see that line 36 is uncovered in the main branch though, so I'll take care of that, but you might have to do a small rebase)

Comment (brucebolt) [reply lib/content_item_loader.rb:54]: Looks like I never pushed the branch after adding the test for the header being nil 🤦‍♂️.

**Review by KludgeKML — APPROVED** [review]
LGTM!

File: lib/content_item_loader.rb (hunk 5 above)
Comment (KludgeKML) [reply lib/content_item_loader.rb:54]: Thanks!
